cpp/src/Slice/MetadataValidator.cpp

```diff
@@ -0,0 +1,410 @@
+    if (info.validArgumentValues.has_value())
+    {
+        // Split the arguments into a string delimited list and check each one.
+        vector<string> argList;
+        IceInternal::splitString(arguments, ",", argList);
+
+        const StringList& validValues = *info.validArgumentValues;
+        for (const auto& arg : argList)
+        {
+            string trimmedArg = IceInternal::trim(arg);
+            if (std::find(validValues.begin(), validValues.end(), trimmedArg) == validValues.end())
+            {
+                string msg = "invalid argument '" + trimmedArg + "' supplied to '" + directive + "' metadata.";
+                p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+                isValid = false;
+            }
+        }
+    }
+
+    // Fourth, we check what the metadata was applied to - does that Slice definition support this metadata?
+    SyntaxTreeBasePtr appliedTo;
+    if (isTypeContext && !info.isTypeMetadata) // This metadata cannot be applied to types, but it was.
+    {
+        string msg = '\'' + directive + "' metadata can only be applied to definitions and declarations" +
+                     ", it cannot be applied directly to types";
+        p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+        isValid = false;
+    }
+    else if (!isTypeContext && !info.isDefinitionMetadata) // This metadata can only be applied to types, but it wasn't.
+    {
+        // Special case: if this metadata can _only_ be applied to types, but it was applied to an operation,
+        // what this really means is that the metadata applies to that operation's return type.
+        // Same thing if this metadata has been applied to a parameter or data member as well.
+        if (auto op = dynamic_pointer_cast<Operation>(p); op && info.isTypeMetadata)
+        {
+            if (const auto returnType = op->returnType())
+            {
+                appliedTo = returnType;
+            }
+            else
+            {
+                string msg = '\'' + directive + "' metadata cannot be applied operations with void return type";
+                p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+                isValid = false;
+            }
+        }
+        else if (auto param = dynamic_pointer_cast<ParamDecl>(p); param && info.isTypeMetadata)
+        {
+            appliedTo = param->type();
+        }
+        else if (auto dm = dynamic_pointer_cast<DataMember>(p); dm && info.isTypeMetadata)
+        {
+            appliedTo = dm->type();
+        }
+        // Otherwise this metadata is just invalid.
+        else
+        {
+            string msg = '\'' + directive + "' metadata cannot be applied to definitions and declarations" +
+                         ", it can only be applied directly to types";
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
```

Comment:
Is this warning required, wouldn't this be handled below by appliedTo checks?

Reply:
I reworked this whole section of checks.